File: app/src/Payment/Application/Queries/GetPayment/GetPaymentHandler.php
<?php

declare(strict_types=1);

namespace Src\Payment\Application\Queries\GetPayment;

use Src\Payment\Application\DTOs\PaymentDTO;
use Src\Payment\Domain\Repositories\PaymentRepositoryInterface;
use Src\Payment\Domain\ValueObjects\PaymentId;
use Src\Shared\Application\Bus\Query;
use Src\Shared\Application\Bus\QueryHandler;
use Src\Shared\Domain\Exceptions\EntityNotFoundException;

/**
 * GET PAYMENT HANDLER (CQRS Pattern)
 * ====================================
 * GetPaymentQuery-ni emal edib PaymentDTO qaytarır.
 *
 * GetOrderHandler ilə eyni yanaşma:
 * 1. Repository-dən entity-ni tap.
 * 2. Entity-ni DTO-ya çevir (domain dəyər xaricə çıxmır).
 * 3. Tapılmadıqda EntityNotFoundException at.
 *
 * EntityNotFoundException vs DomainException:
 * - DomainException: Biznes qaydası pozuldu (422).
 * - EntityNotFoundException: Resurs tapılmadı (404). DomainException-un alt sinfidir.
 * bootstrap/app.php-dəki exception handler fərqli HTTP status kod qaytarır.
 */
class GetPaymentHandler implements QueryHandler
{
    public function __construct(
        private readonly PaymentRepositoryInterface $paymentRepository,
    ) {}

    public function handle(Query $query): PaymentDTO
    {
        /** @var GetPaymentQuery $query */

        $payment = $this->paymentRepository->findById(
            new PaymentId($query->paymentId()),
        );

        if ($payment === null) {
            throw new EntityNotFoundException("Ödəniş tapılmadı: {$query->paymentId()}");
        }

        return PaymentDTO::fromEntity($payment);
    }
}

File: app/spring/src/main/java/az/ecommerce/payment/application/query/getpayment/GetPaymentHandler.java
package az.ecommerce.payment.application.query.getpayment;

import az.ecommerce.payment.application.dto.PaymentDto;
import az.ecommerce.payment.domain.repository.PaymentRepository;
import az.ecommerce.payment.domain.valueobject.PaymentId;
import az.ecommerce.shared.application.bus.QueryHandler;
import az.ecommerce.shared.domain.exception.EntityNotFoundException;
import org.springframework.stereotype.Service;

@Service
public class GetPaymentHandler implements QueryHandler<GetPaymentQuery, PaymentDto> {
    private final PaymentRepository repository;
    public GetPaymentHandler(PaymentRepository r) { this.repository = r; }
    @Override public PaymentDto handle(GetPaymentQuery q) {
        return repository.findById(new PaymentId(q.paymentId()))
                .map(PaymentDto::fromDomain)
                .orElseThrow(() -> new EntityNotFoundException("Payment", q.paymentId().toString()));
    }
}
